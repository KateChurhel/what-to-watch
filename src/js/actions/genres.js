/* istanbul ignore file */
import { ADD_ACTIVE_GENRE, REMOVE_ACTIVE_GENRE, CLEAR_ACTIVE_GENRES } from '../constants/actions/genres';

export const addActiveGenre = (id) => ({
  type: ADD_ACTIVE_GENRE,
  id,
});

export const removeActiveGenre = (id) => ({
  type: REMOVE_ACTIVE_GENRE,
  id,
});

export const clearActiveGenres = () => ({
  type: CLEAR_ACTIVE_GENRES,
});
