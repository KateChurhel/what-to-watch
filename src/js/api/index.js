// helpers
import { RequestError } from '../helpers/exceptions';

const baseUrl = 'https://api.themoviedb.org/3';
const apiKey = '6b19f7c4abb404c76cdab04bab4aaa89';

const contentTypes = {
  json: 'application/json',
  // text: 'text/html',
};

const encodeQueryData = (data) => {
  if (!data) {
    return '';
  }

  try {
    const queryStringStart = '?';

    return Object.keys(data).reduce((queryString, key, index) => {
      const paramStringStart = index ? '&' : '';
      let paramString = '';
      if (encodeURIComponent(data[key])) {
        paramString = `${paramStringStart}${encodeURIComponent(key)}=${encodeURIComponent(data[key])}`;
      }

      return queryString + paramString;
    }, queryStringStart);
  } catch (e) {
    throw new TypeError('Error: incorrect query data format.');
  }
};

const getResponseData = async (response) => {

  const contentType = response.headers.get('content-type');

  if (contentType.includes(contentTypes.json)) {
    return response.json();
  }

  throw new TypeError('Error: incorrect received data format.');
};

const sendRequest = async (url = '', options = {}) => {
  const requestUrl = baseUrl + url;
  const requestHeaders = new Headers({
    'Content-Type': contentTypes.json,
  });

  const requestOptions = {
    headers: requestHeaders,
    ...options,
  };
  const response = await fetch(requestUrl, requestOptions);
  const responseData = await getResponseData(response);

  if (!response.ok) {
    throw new RequestError(response.status, responseData);
  }

  return responseData;
};

export const sendGetRequest = (url, data) => {
  const requestUrl = url + encodeQueryData({ ...data, api_key: apiKey, language: 'en-US' });
  const requestOptions = {
    method: 'GET',
  };

  return sendRequest(requestUrl, requestOptions);
};

export default sendGetRequest;
