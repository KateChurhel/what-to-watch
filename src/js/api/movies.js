/* istanbul ignore file */
import { sendGetRequest } from './index';

export const getCategoryData = (category, data) => sendGetRequest(`/${category}/popular`, data);

export const getDetailsData = (category, id, data) => sendGetRequest(`/${category}/${id}`, data);

export const getSearchResult = (category, data) => sendGetRequest(`/search/${category}`, data);

export const getDiscoverResult = (data) => sendGetRequest('/discover/movie', data);

export const getGenres = () => sendGetRequest('/genre/movie/list');
