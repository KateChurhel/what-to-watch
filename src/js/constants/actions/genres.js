export const ADD_ACTIVE_GENRE = 'activeGenres/add';
export const REMOVE_ACTIVE_GENRE = 'activeGenres/remove';
export const CLEAR_ACTIVE_GENRES = 'activeGenres/clear';