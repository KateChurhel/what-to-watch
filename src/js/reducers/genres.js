import { ADD_ACTIVE_GENRE, REMOVE_ACTIVE_GENRE, CLEAR_ACTIVE_GENRES } from '../constants/actions/genres';

const initialState = {
  activeGenres: [],
};

const genres = (state = initialState, action) => {
  const activeGenres = state.activeGenres.filter((genre) => genre !== action.id);

  switch (action.type) {
    case ADD_ACTIVE_GENRE:
      return {
        activeGenres: [...state.activeGenres, action.id],
      };
    case REMOVE_ACTIVE_GENRE:
      return { activeGenres };
    case CLEAR_ACTIVE_GENRES:
      return {
        activeGenres: [],
      };
    default:
      return state;
  }
};

export default genres;
