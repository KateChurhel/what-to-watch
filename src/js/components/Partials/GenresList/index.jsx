// libraries
import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useHistory, useParams } from 'react-router-dom';
// api
import { getGenres } from '../../../api/movies';
// actions
import { addActiveGenre, removeActiveGenre, clearActiveGenres } from '../../../actions/genres';

const GenresList = () => {
  const [genresData, setGenresData] = useState([]);
  const dispatch = useDispatch();
  const history = useHistory();
  const { category } = useParams();
  const activeGenres = useSelector(({ genres }) => genres.activeGenres);

  const getGenresData = async () => {
    const { genres = [] } = await getGenres();
    setGenresData([{ id: 0, name: 'All Genres' }, ...genres]);
  };

  useEffect(() => {
    getGenresData();
  }, []);

  return genresData.length ? (
    <ul className="catalog__genres-list">
      {genresData.map(({ id, name }) => {
        const isActive = (id || activeGenres.length) ? activeGenres.includes(id) : true;

        return (
          <li key={id} className={`catalog__genres-item ${isActive ? 'catalog__genres-item--active' : ''}`}>
            <Link
              className="catalog__genres-link"
              to="/"
              onClick={(event) => {
                event.preventDefault();
                if (id) {
                  if (isActive) {
                    dispatch(removeActiveGenre(id));
                  } else {
                    dispatch(addActiveGenre(id));
                  }
                } else {
                  dispatch(clearActiveGenres());
                }
                history.push(`/${category}/1/`);
              }}
            >
              {name}
            </Link>
          </li>
        );
      })}
    </ul>
  ) : null;
};

export default GenresList;
