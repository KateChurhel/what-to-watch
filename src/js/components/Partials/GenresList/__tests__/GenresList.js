import React from 'react';
import { act, render } from '@testing-library/react';
import { useSelector } from 'react-redux';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { waitForElement } from '@testing-library/dom';
import GenresList from '../index';
import { getGenres } from '../../../../api/movies';
import { MOCK_GENRES } from '../../../../mocks/movies';
import { addActiveGenre, removeActiveGenre, clearActiveGenres } from '../../../../actions/genres';
import CATEGORIES from '../../../../constants/categories';
import ROUTES from '../../../../constants/routes';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => jest.fn(),
}));

jest.mock('../../../../api/movies');
jest.mock('../../../../actions/genres', () => ({
  addActiveGenre: jest.fn(),
  removeActiveGenre: jest.fn(),
  clearActiveGenres: jest.fn(),
}));

describe('Genres List: ', () => {
  let wrapper;

  beforeAll(() => {
    getGenres.mockReturnValue({ genres: MOCK_GENRES });
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('correctly renders', async () => {
    useSelector.mockImplementation((selectorFn) => selectorFn({ genres: { activeGenres: [] } }));

    const { asFragment, getByText } = render(
      <MemoryRouter initialEntries={[`/${CATEGORIES.movie}/1`]} initialIndex={1}>
        <Route path={ROUTES.catalog}>
          <GenresList />
        </Route>
      </MemoryRouter>,
    );

    await waitForElement(() => getByText('All Genres'));

    expect(asFragment()).toMatchSnapshot();
  });

  it('correctly renders with active value "All Genres"', async () => {
    useSelector.mockImplementation((selectorFn) => selectorFn({ genres: { activeGenres: [] } }));

    await act(async () => {
      wrapper = await mount(
        <MemoryRouter initialEntries={[`/${CATEGORIES.movie}/1`]} initialIndex={1}>
          <Route path={ROUTES.catalog}>
            <GenresList />
          </Route>
        </MemoryRouter>,
      );
    });
    wrapper.update();

    expect(wrapper.find('.catalog__genres-item--active .catalog__genres-link').text()).toEqual('All Genres');
  });

  it('correctly renders with active filter value', async () => {
    useSelector.mockImplementation((selectorFn) => selectorFn({ genres: { activeGenres: [MOCK_GENRES[0].id] } }));

    await act(async () => {
      wrapper = await mount(
        <MemoryRouter initialEntries={[`/${CATEGORIES.movie}/1`]} initialIndex={1}>
          <Route path={ROUTES.catalog}>
            <GenresList />
          </Route>
        </MemoryRouter>,
      );
    });
    wrapper.update();

    expect(wrapper.find('.catalog__genres-item--active .catalog__genres-link').text()).toEqual(MOCK_GENRES[0].name);
  });

  it('correctly renders with several active filter values', async () => {
    const activeGenres = [MOCK_GENRES[0].id, MOCK_GENRES[1].id];
    useSelector.mockImplementation((selectorFn) => selectorFn({ genres: { activeGenres } }));

    await act(async () => {
      wrapper = await mount(
        <MemoryRouter initialEntries={[`/${CATEGORIES.movie}/1`]} initialIndex={1}>
          <Route path={ROUTES.catalog}>
            <GenresList />
          </Route>
        </MemoryRouter>,
      );
    });
    wrapper.update();

    expect(wrapper.find('.catalog__genres-item--active .catalog__genres-link').length).toEqual(activeGenres.length);
  });

  describe('', () => {
    beforeEach(async () => {
      useSelector.mockImplementation((selectorFn) => selectorFn({ genres: { activeGenres: [MOCK_GENRES[0].id] } }));

      await act(async () => {
        wrapper = await mount(
          <MemoryRouter initialEntries={[`/${CATEGORIES.movie}/1`]} initialIndex={1}>
            <Route path={ROUTES.catalog}>
              <GenresList />
            </Route>
          </MemoryRouter>,
        );
      });
      wrapper.update();
    });

    it('click on inactive link', async () => {
      const link = wrapper.findWhere((node) => (
        node.name() === 'a'
          && node.text() === MOCK_GENRES[2].name
      ));
      link.simulate('click');
      expect(addActiveGenre).toHaveBeenCalled();
    });

    it('click on active link', async () => {
      const link = wrapper.find('.catalog__genres-item--active .catalog__genres-link');
      link.simulate('click');
      expect(removeActiveGenre).toHaveBeenCalled();
    });

    it('click on "All Genres" link', async () => {
      const link = wrapper.findWhere((node) => (
        node.name() === 'a'
          && node.text() === 'All Genres'
      ));
      link.simulate('click');
      expect(clearActiveGenres).toHaveBeenCalled();
    });
  });
});
