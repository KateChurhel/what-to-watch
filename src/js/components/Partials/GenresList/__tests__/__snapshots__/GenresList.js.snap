// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Genres List:  correctly renders 1`] = `
<DocumentFragment>
  <ul
    class="catalog__genres-list"
  >
    <li
      class="catalog__genres-item catalog__genres-item--active"
    >
      <a
        class="catalog__genres-link"
        href="/"
      >
        All Genres
      </a>
    </li>
    <li
      class="catalog__genres-item "
    >
      <a
        class="catalog__genres-link"
        href="/"
      >
        Fantasy
      </a>
    </li>
    <li
      class="catalog__genres-item "
    >
      <a
        class="catalog__genres-link"
        href="/"
      >
        Horror
      </a>
    </li>
    <li
      class="catalog__genres-item "
    >
      <a
        class="catalog__genres-link"
        href="/"
      >
        Music
      </a>
    </li>
    <li
      class="catalog__genres-item "
    >
      <a
        class="catalog__genres-link"
        href="/"
      >
        Mystery
      </a>
    </li>
    <li
      class="catalog__genres-item "
    >
      <a
        class="catalog__genres-link"
        href="/"
      >
        Romance
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
